### ISSUE TITLE
Datachannel onclose event does not trigger

### ISSUE THREAD

**Issue by mafrost** [node-webrtc/node-webrtc#236]
I do not get the onclose event from datachannels. I've tested this with the latest 4 versions and both on Linux (Ubuntu 14.04) and OS X (Maverick). I've tested with both the example code and custom code but the onclose event is not triggered. Have I missed something here?

**Comment by mafrost** [comment]
It seems I might have misunderstood how the onclose event is triggered. I thought it should be triggered when a client disconnects (closes a tab) but it seems it only triggers if .close() is called on the channel?

**Comment by melv-n** [comment]
@mafrost It should **always** trigger when a channel is closed, even if it is closed abruptly. You can see this behaviour in the browser when a connection is closed. This is broken/not implemented in the current version of `node-ipc`. I hope they will take this into account and find a solution to this as soon as possible.

@markandrus  ping

**Comment by melv-n** [comment]
Not firing this event might be part of the reason DataChannels appear to not close correctly in some situations, resulting in SegFaults during Garbage Collection. Thoughts on this?

**Comment by strbean** [comment]
Bump, I've been testing this behavior locally. Oddly enough, refreshing the tab seems to break things far worse than closing it. When I refresh the tab, suddenly the PeerConnection and DataChannel associated with that tab become poison. Even accessing DataChannel.readyState causes a segfault.

**Comment by vikeshkhanna** [comment]
Bump, was anyone able to solve this issue?

**Comment by nazar-pc** [comment]
I've added one more stack trace in #325, can someone take a look at it?
I think this is a major issue for any kind of use that is even close to production (my CI fails too often because of this).

**Comment by nazar-pc** [comment]
Here is probably the simplest possible reduced test case to reproduce this issue reliably (100%):
```javascript
var wrtc = require('wrtc');

var peer1       = new wrtc.RTCPeerConnection({iceServers : []});
var peer2       = new wrtc.RTCPeerConnection({iceServers : []});
var channel1    = peer1.createDataChannel('data', {negotiated : true, id : 0});
var channel2    = peer2.createDataChannel('data2', {negotiated : true, id : 0});
var waiting_for = 2;

function ready () {
	--waiting_for;
	if (!waiting_for) {
		channel2.close();
		peer2.close();
		setTimeout(function () {
			channel1.send('Hello');
		});
	}
}

channel1.onopen = ready;
channel2.onopen = ready;

peer1.createOffer()
	.then(function (offer) {
		return peer1.setLocalDescription(offer);
	})
	.then(function () {
		peer2.setRemoteDescription(peer1.localDescription);
		return peer2.createAnswer();
	})
	.then(function (answer) {
		return peer2.setLocalDescription(answer);
	})
	.then(function () {
		peer1.setRemoteDescription(peer2.localDescription);
	});
```

Hopefully this will help with fixing this issue.

Please node that without `setTimeout()` it doesn't crash.

**Comment by markandrus** [comment]
Thanks! I’m on vacation, but will write up a test and a fix for this soon (unless someone else wants to try their hand in the meantime).

**Comment by nazar-pc** [comment]
`close` event is actually triggered already, the issue was in different part.
I've filled a fix in #342, which also contains corresponding test case.
